### PR TITLE
Toggle details refactor

### PR DESCRIPTION
**Pull request by ka1130**
This PR is a refactor of [toggling details PR#7](https://github.com/ka1130/TVloVer/pull/7) combined with [PR#8](https://github.com/ka1130/TVloVer/pull/8).

 - Details are shown once any show is clicked (clickability is indicated by a hover effect) in a modal
 - Details include show's title, image (placeholder in the case of lacking image), summary, formatted date and time
 - The visibility of modal is handled by internal App's state, it's being closed by either clicking the close button or clicking outside of the modal
 - Pagination is handled by Redux, there's also an indicator for active page

**Screencast**:
peek 2018-10-09 23-31

### CODE REVIEW

File: src/components/App/index.js [hunk 1]
```diff
@@ -1,39 +1,76 @@
...
+
+  openModal = (episode) => {
+    this.setState({ 
+      isModalVisible: true,
+      activeEpisode: episode
+    });
+  }
+
+  closeModal = event => {
+    const { id } = event.target;
```
Comment (yakato) [inline src/components/App/index.js]: you don't need any ide or event here, just set state of parent on click in child

File: src/components/App/index.js [hunk 2]
```diff
@@ -1,39 +1,76 @@
...
+    if (id === "modalBg" || id === "closeModal") this.setState({ isModalVisible: false })
   }
 
   render() {
-    const { error, loading, episodes } = this.props;
+    const { error, loading, episodes, currentPage, episodesPerPage } = this.props;
+    const {isModalVisible, activeEpisode } = this.state;
+    const indexOfLastEpisode = parseInt(currentPage) * episodesPerPage;
```
Comment (yakato) [inline src/components/App/index.js]: logic for pagination maybe can be extract somewhere? consider use `reselect` lib

File: src/redux/reducers/index.js [hunk 3]
```diff
@@ -1,8 +1,14 @@
 import { combineReducers } from 'redux';
 import data from 'redux/reducers/episodesReducer';
+import getCurrentPage from 'redux/reducers/paginationReducer';
 
 const rootReducer = combineReducers({
-  data
+  data,
+  getCurrentPage,
+  episodesPerPage: 12,
```
Comment (yakato) [inline src/redux/reducers/index.js]: it is still for episodes so should be inside episodesReducer
reducer for pagination shouldn't be named same as action, you can rename it into `pagination` for example or just `currentPage` key

File: src/redux/reducers/paginationReducer.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  currentPage: 1,
+  episodesPerPage: 12
+};
+
+export default function getCurrentPage(state = initialState, action) {
+  switch(action.type) {
+    case GET_CURRENT_PAGE:
+      return {
+        currentPage: action.payload,
+        episodesPerPage: 12
```
Comment (yakato) [inline src/redux/reducers/paginationReducer.js]: how the `episodesPerPage` can be change? on what type of action? Maybe until it's hardcoded value (not coming from API and it's not selected from API data) consider to pass it into component directly instead push it to the store

File: src/components/App/index.js (hunk 1 above)
Comment (ka1130) [reply src/components/App/index.js]: I didn't manage to handle this without id, it ended up with modal being closed on click anywhere. :/

File: src/redux/reducers/paginationReducer.js (hunk 4 above)
Comment (ka1130) [reply src/redux/reducers/paginationReducer.js]: Right, moved directly to the `<App />` component.

File: src/components/App/Pagination/index.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import styles from './Pagination.module.scss';
+
+class Pagination extends Component {
+  renderPageNumbers = () => {
+    const { episodes, currentPage, getCurrentPage, episodesPerPage } = this.props;
+
+    let pageNumbers = [];
+    let i = 1;
+    while(i <= Math.ceil(episodes.length / episodesPerPage)) pageNumbers.push(i++);
```
[the remark this answers is not in the document]
Comment (ka1130) [reply src/components/App/Pagination/index.js]: I tried and failed to see how to do it when I only accumulate elements in an array, without doing anything else with them. Maybe I can try reduce on another occasion? :)

File: src/components/App/index.js (hunk 2 above)
Comment (ka1130) [reply src/components/App/index.js]: Good idea. :) I used it, for the first time and hope it's fine the way I handled it. I needed currentEpisodes on the top level component together with other state elements but I assigned it a result of selector's computing. Please let me know what do you think about it.

**Comment by ka1130** [comment]
Thanks for your review, @yakato! As always lots of great stuff to learn, definitely `reselect` is one of them + a recap of ES6 array methods and usage of the `classnames` lib, I'll keep that in mind. Please have a second look at the  changes introduced. :)

File: src/redux/actions/paginationActions.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { GET_CURRENT_PAGE } from 'redux/actions/types';
+
+export const getCurrentPage = (page = 1) => {
```
Comment (yakato) [inline src/redux/actions/paginationActions.js]: Don't you think that this action actually setting current page not getting?

File: src/redux/reducers/index.js [hunk 7]
```diff
@@ -1,8 +1,13 @@
 import { combineReducers } from 'redux';
 import data from 'redux/reducers/episodesReducer';
+import getCurrentPage from 'redux/reducers/paginationReducer';
 
 const rootReducer = combineReducers({
-  data
+  data,
+  getCurrentPage,
```
Comment (yakato) [inline src/redux/reducers/index.js]: please name it `currentPage`, names of keys of the store should be like a database keys

Comment (yakato) [reply src/redux/reducers/index.js]: Now your store looks weird with this key
screen shot 2018-10-12 at 16 02 58

File: src/components/App/EpisodesList/Show/index.js [hunk 8]
```diff
@@ -25,11 +25,11 @@ class Show extends Component {
   render() {
```
[the remark this answers is not in the document]
Comment (ka1130) [reply src/components/App/EpisodesList/Show/index.js]: Yeah, forgot about this one, my bad. Renamed. :)

File: src/components/App/index.js [hunk 9]
```diff
@@ -1,39 +1,78 @@
...
+  closeModal = event => {
+    const { id } = event.target;
+    if (id === "modalBg" || id === "closeModal") this.setState({ isModalVisible: false })
   }
 
   render() {
-    const { error, loading, episodes } = this.props;
+    const { error, loading, currentEpisodes } = this.props;
+    const { episodesPerPage } = this.state;
+    const { isModalVisible, activeEpisode } = this.state;
```
[the remark this answers is not in the document]
Comment (ka1130) [reply src/components/App/index.js]: Right! Correcting.

File: src/redux/actions/paginationActions.js (hunk 6 above)
Comment (ka1130) [reply src/redux/actions/paginationActions.js]: True. :) I'm changing the name to a more appropriate.

File: src/redux/reducers/index.js (hunk 7 above)
Comment (ka1130) [reply src/redux/reducers/index.js]: Right, I've changed the name.

File: src/redux/selectors/paginationSelector.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { createSelector } from 'reselect';
+
+const episodesPerPage = 12;
+
+export default createSelector(
+  state => state.data.episodes,
+  state => state.getCurrentPage.currentPage,
+  (episodesState, currentPageState) => {
+    const indexOfLastEpisode = parseInt(currentPageState) * episodesPerPage;
+    const indexOfFirstEpisode = indexOfLastEpisode - episodesPerPage;
```
[the remark this answers is not in the document]
Comment (ka1130) [reply src/redux/selectors/paginationSelector.js]: :smiley:

**Comment by ka1130** [comment]
Thanks for all your input, @yakato :) I've just introduced all the improvements + the forgotten rename of the `Show` component to `Episode`. Merging! :tada:

**Pull request merged**